### PR TITLE
cleanup of Filesystem-related functions

### PR DESCRIPTION
**Pull request by vtjnash**
1) groups Filesystem-related functions into a new module "Filesystem" (was "Base.FS")
2) deprecates isreadable/writable/executable for filesystem paths, closing #7385
3) adds a few missing `isopen` tests

### CODE REVIEW

**Comment by tkelman** [comment]
capitalization bikeshed - `Filesystem` or `FileSystem` ?

**Comment by dpsanders** [comment]
+1 for FileSystem

**Comment by vtjnash** [comment]
filesystem is a word

**Comment by StefanKarpinski** [comment]
https://en.wikipedia.org/wiki/File_system: can be two words or one.

**Comment by StefanKarpinski** [comment]
More prevalent usage seems to be two words; frequent abbreviation as "FS" seems to fit well with calling it FileSystem.

**Comment by stevengj** [comment]
+1 for "Filesystem".  Since the one-word version is already standard English, using CamelCase to concatenate the two-word version looks unnatural to me.

**Comment by ScottPJones** [comment]
I don't care either way, just data points: Apache Hadoop API -> `FileSystem`, Java -> `FileSystem`.

**Comment by StefanKarpinski** [comment]
Ruby Zip module: http://www.rubydoc.info/gems/rubyzip/Zip/FileSystem

**Comment by stevengj** [comment]
There is [Boost.Filesystem](http://www.boost.org/doc/libs/1_59_0/libs/filesystem/doc/index.htm).  On the other hand, there is the Microsoft Visual Basic [FileSystem module](https://msdn.microsoft.com/en-us/library/2f66wza0(v=vs.90).aspx).  But I don't have strong feelings either way.

**Comment by vtjnash** [comment]
Ruby seems pretty inconsistent there, since they use `FileSystem`, but `Fs`

**Comment by ScottPJones** [comment]
Don't care, either way is better than just FS.

**Comment by StefanKarpinski** [comment]
Since FileSystem seems most common and "file system" is a valid spelling, let's go with that.

**Comment by vtjnash** [comment]
that seems to be a lot of emphasis on "System", which I don't think is warranted given that I rarely see "file system" in a document without also finding "filesystem" in the same document somewhere else (indicating that this has rapidly become a word). I'm not really interested in bikeshedding this, but if someone wants to propose a shorter name, I might be listening. Python uses `os.path`. I would be OK going with `Files`.

**Comment by quinnj** [comment]
What about `Bash` or `Shell`? I think it'd be cool to do `Shell.cd` and `Shell.touch`. `Files` isn't too bad either, but I think I'd lean towards `FileSystem` over `Files`, despite the brevity.

**Comment by vtjnash** [comment]
This doesn't use the shell or interact with it in any way. I still just don't like the way FileSystem seems to emphasize the word "System", since filesystem is already a valid regular word.

**Comment by tkelman** [comment]
No need to rush this, I don't see why this would need to be an 0.4 target. `iswritable` is a more convenient name and easier to remember than fiddling with octal filemodes, and the user inaccuracy issue is a pretty confusing reason for deprecation.

**Comment by ScottPJones** [comment]
A file system contains directories and files, and doesn't this API do more than just deal with files?
(which might imply name should be `filesystem` or `FileSystem`)
`cd` is changing process or shell state, not changing anything in a file system, so I wouldn't put it under `FileSystem`.

**Comment by StefanKarpinski** [comment]
This is absolutely not going into 0.4.

**Comment by vtjnash** [comment]
rebased to be ready to merge when CI is happy

**Comment by tkelman** [comment]
deprecation warnings should actually tell the user what replacement code to use.

**Comment by vtjnash** [comment]
> deprecation warnings should actually tell the user what replacement code to use.

i can't really predict why the user thought they wanted to call these functions, so I pointed them to the man page where it describes why linux thinks of them as deprecated.

**Comment by tkelman** [comment]
What are they documented to do? Deprecated functions should have something done about their docs, but not sure if that should happen at deprecation time or when the deprecation gets removed.

The deprecation warning is not the right place to send someone to read a man page because they called a function that you don't like. People are also using this function on systems that don't have such a man page. Just suggest the replacement code `(st.mode & 0o222) > 0` in the warning.

I'm still not convinced there's a good reason to get rid of these since the replacement is inconvenient, low-level and unix-specific code.

File: base/deprecated.jl [hunk 1]
```diff
@@ -786,4 +786,25 @@ export FloatingPoint
...
+function isreadable(st::FS.StatStruct)
+    depwarn("isreadable is deprecated as it implied that the file would actually be readable by the user. see also the man page for `access`", :isreadable)
+    return (st.mode & 0o444) > 0
+end
+function iswritable(st::FS.StatStruct)
+    depwarn("iswritable is deprecated as it implied that the file would actually be writable by the user. see also the man page for `access`", :iswritable)
+    return (st.mode & 0o222) > 0
+end
+function isexecutable(st::FS.StatStruct)
+    depwarn("isexecutable is deprecated as it implied that the file would actually be executable by the user. see also the man page for `access`", :isreadable)
```
Comment (tkelman) [inline base/deprecated.jl]: `:isexecutable`

File: base/deprecated.jl [hunk 2]
```diff
@@ -786,4 +786,25 @@ export FloatingPoint
     Regex(pattern, flags)
 end
 
+# Filesystem module updates
```
Comment (tkelman) [inline base/deprecated.jl]: These should be 0.5 deprecations now

**Comment by vtjnash** [comment]
> What are they documented to do? ... ok I think I get the issue now, the past behavior was not specific enough?

what are they supposed to do? currently they are defined to return true if there might be a user (excluding root) that could have done the specified action at the instant prior to the time the question was answered. however, i can't think of a place case where that definition is useful.

the current docs say:
```
Returns `true` if the current user has permission to read `path`, `false` otherwise.
```
but that is not a question that `stat` can answer, as discussed in https://github.com/JuliaLang/julia/issues/7385

**Comment by tkelman** [comment]
We don't usually give reasons for deprecation in warnings, we primarily tell the user what to change their code to. Here the ambiguity means users who care probably should do some nontrivial refactoring, but we should at least spell out in the warning how to accomplish what these functions used to do (even if technically buggy).

**Comment by vtjnash** [comment]
OK, but: *what were they supposed to do?*

**Comment by simonster** [comment]
Most likely what the docs said they did?

**Comment by vtjnash** [comment]
they didn't do what the docs said they did. so is that a vote for fixing it to do what the docs said it did (and then deprecating that because it uses a deprecated syscall), or is that a vote for fixing the docs to tell you that it doesn't do what you expect (then deprecating it because having a function that doesn't do anything useful is pretty pointless)?

**Comment by simonster** [comment]
It's a vote for making the deprecation tell you how to do what they said they did (assuming it's actually possible to do).

**Comment by vtjnash** [comment]
the closest option would be to reinstate @ihnorton's PR (https://github.com/JuliaLang/julia/pull/7418/files). i'm not entirely clear on why he reverted that.

**Comment by ihnorton** [comment]
I just grep'd through a ~2 week old clone of all packages in METADATA. As far as I can tell, there are only two uses of `isreadable(path)`: @aviks JavaCall, and @hng's BiomolecularStructures.jl. I don't see any uses of `iswritable(path)`. Based on the preceding notification to them, and extrapolating such low usage numbers to the broader ecosystem, I'm +1 for removing these functions without deprecation because the documentation was so wildly wrong.

I would suggest something like `path_has_permissions(path, "ugo+rwx")` or similar as a replacement for the current *actual* definition of `isreadable` and friends (i.e. non-octal convenience functions), but the whether/how discussion can happen elsewhere.

**Comment by vtjnash** [comment]
thanks for checking. it looks like both cases could be testing `isfile` (or `isdir` or `ispath`), although JavaCall might be able to use `Libdl.find_library` instead to simplify the code there?

**Comment by tkelman** [comment]
That is encouraging. I feel like I've seen `isexecutable` a few more places than that.

**Comment by ihnorton** [comment]
Oh, right: for `isexecutable` it is JavaCall (@aviks again), Homebew (@staticfloat), MATLAB (JuliaLang).

**Comment by vtjnash** [comment]
looks like the same answer would be good: use `isfile` instead

**Comment by aviks** [comment]
`Libdl.find_library()` is exactly what I need in place of isreadable. 

I suppose I could just use `isfile()` instead of `isexecutable`. 

So consider this an ack and a +1

**Comment by staticfloat** [comment]
It's not really necessary in my case to check for the permissions bits, so I changed to `isfile()`.

**Comment by vtjnash** [comment]
bump. updated for 0.5. are there any last comments before this gets merged?

**Comment by yuyichao** [comment]
Make the `FS` binding deprecated?

**Comment by tkelman** [comment]
And make the deprecations actually suggest replacement code. That's still not done here.

**Comment by hayd** [comment]
Could FS also become a package/removed from base? (I'm guessing no, but thought I'd mention it.)

**Comment by vtjnash** [comment]
it mostly could (although basic infrastructure like `include` needs it), and might be moved into more of a "preinstalled package". Unlike many of the other packages, I don't see a strong likelihood of anyone considering developing a competing external implementation of this module.

**Comment by tkelman** [comment]
Though it would be cool to selectively sandbox code to disable running anything in this module.

**Comment by vtjnash** [comment]
disabling this code doesn't result in a sandbox, it just makes makes coding more annoying by needing to copy the code locally

**Comment by vtjnash** [comment]
updated to address comments

File: test/socket.jl [hunk 3]
```diff
@@ -87,7 +87,7 @@ wait(port)
 wait(tsk)
 
 socketname = (@windows ? "\\\\.\\pipe\\uv-test" : "testsocket") * "-" * randstring(6)
-@unix_only isfile(socketname) && Base.FS.unlink(socketname)
+@unix_only isfile(socketname) && Base.Filesystemunlink(socketname)
```
Comment (tkelman) [inline test/socket.jl]: missing a `.`

File: base/docs/helpdb.jl [hunk 4]
```diff
@@ -2024,13 +2024,6 @@ Reseed the random number generator. If a `seed` is provided, the RNG will give a
 srand
 
 doc"""
-    isexecutable(path) -> Bool
-
-Returns `true` if the current user has permission to execute `path`, `false` otherwise.
```
Comment (tkelman) [inline base/docs/helpdb.jl]: should remove from rst if you're removing this docstring entirely. also regen the rst for the adjustments to isreadable and iswritable docstrings.

Comment (vtjnash) [inline base/docs/helpdb.jl]: thanks, i didn't realize that was how it worked

File: test/socket.jl (hunk 3 above)
Comment (vtjnash) [inline test/socket.jl]: this line disappears in the rebase conflict resolution anyways, but thank

File: base/docs/helpdb.jl (hunk 4 above)
Comment (tkelman) [inline base/docs/helpdb.jl]: I think it's most clearly documented at https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#improving-documentation right now, but yes kind of confusing at the moment.

+1 to merge when green.

**Pull request merged**